File: string/di-string-match.py
# https://leetcode.com/problems/di-string-match/
# tags - string
class Solution:
    def diStringMatch(self, s: str) -> List[int]:
        perm = []
        left, right = 0, len(s)
        for char in s:
            if char == 'I':
                # Pick the smallest integer left.
                # It is guaranteed to be smaller than the next integer.
                perm.append(left)
                left += 1
            else:
                # Pick the largest integer left.
                perm.append(right)
                right -= 1
        perm.append(left)
        return perm
            

File: string/di-string-match.js
// https://leetcode.com/problems/di-string-match/
// tags - string
/**
 * @param {string} s
 * @return {number[]}
 */
var diStringMatch = function (s) {
  const n = s.length;
  const result = [];
  let low = 0,
    high = n;
  for (let i = 0; i < s.length; i++) {
    if (s[i] === "I") {
      // Next number is guaranteed to be greater.
      result.push(low);
      low++;
    } else {
      // Next number is guaranteed to be smaller.
      result.push(high);
      high--;
    }
  }
  // Can push either high/low - that'll be the last remaining number.
  result.push(low);

  return result;
};
